### DISCUSSION TITLE
Debouncing doesn't seem to be implemented in a useful way

### DISCUSSION THREAD

**Discussion by diesieben07** [OptimalBits/bull#2773] [Q&A] [answered]
The relatively new `debounce` option on jobs seems to cover my use-case exactly. I want to run an export job whenever a piece of data changes. When the data changes rapidly, I want to only export it _once_.

However, during the "debounce TTL" new jobs with the same debounce ID will be ignored entirely. This causes the following problem:

- Change Data from A to B, job triggers _immediately_ ("B" is exported)
- Change Data from B to C within debounce TTL, job does not trigger at all

Now the exported data is out of date.

What I would expect to happen is:
- Change Data from A to B, job is _pending_ (delayed by debounce TTL)
- Change Data from B to C within debounce TTL, pending Job's delay is updated so it now runs [debounce TTL] after _this_ change
- Eventually debounce TTL passes _after the last change_, job runs, _latest data is exported_.

How to achieve this desired behavior?

**Comment by roggervalf** [comment]
hi @diesieben07, you also can use delay along side with your ttl in this case. Same delay value as your ttl if you don't want to have eventual consistency.

**Reply by diesieben07** [reply]
Thank you for your reply, however I am a bit confused. I think what you mean is something like setting debounce-ttl to 10_000 and delay to 11_000. Then for the 10_000 no new tasks will be scheduled, but at 11_000 the delayed one will be ran.
If I understand this right, this will be sort of like a throttling (at most one every 10_000ms), right? I think that would be acceptable, even though it is not what I was originally planning to have.

> Same delay value as your ttl if you don't want to have eventual consistency.

This also confused me a bit, I specifically _do_ want eventual consistency. Once the mutations have settled, I want the export to run again so that the final, settled data is exported.

**Reply by roggervalf** [reply, marked as answer]
hey, so there are some cases to consider:
* debounce ttl without delay i.e 10_000, means that any new job with same debounce id will be ignore in that period. Your first job that was added in this period will be processed as soon as one worker picked it. If this job completes before this period, for the remaining time, you will ignore other jobs with same debounce id.
* debounce ttl with delay (same value) i.e 10_000, means that this first job will be added as a delayed one. In this period other jobs will be ignored, but after that time job will be move to wait and be ready to be processed by a worker. Also while this job is in wait state or active, another job with same debounce id can be added (you can always add these jobs with same delay and ttl config).
* In case you want also to ignore jobs while this first job is in active state (not completed), use extended mode (no ttl).

**Reply by diesieben07** [reply]
Thank you for the clarification. I understood it correctly then, debounce ttl plus a delay effectively becomes throttle (run the job at most once every 10 seconds). This works for me, however it isn't truly what I wanted.

What I wanted was that the job _does not run_ while the mutations are still coming in, instead it runs _once_ at the end.
